morango/models/core.py

```diff
@@ -572,97 +582,146 @@ class Meta:
+                .distinct()
+            )
+            queryset = queryset.filter(instance_id__in=matching_store_instances)
+        return dict(queryset.values_list("instance_id", "counter"))
 
-        queryset = cls.objects.all()
+    @classmethod
+    def calculate_filter_specific_instance_counters(
+        cls,
+        filters,
+        is_producer=False,
+        v2_format=False,
+        only_if_in_store=True,
+    ):
 
-        per_filter_max = []
+        if v2_format:
 
-        for filt in filters:
-            # {filt} LIKE partition || '%'
-            qs = queryset.annotate(
-                filter_matches=ValueStartsWithField(filt, "partition")
+            queryset = cls.objects.all()
+
+            # get the DMC records with partitions that fall under the filter prefixes
+            sub_condition = functools.reduce(
+                lambda x, y: x | y,
+                [Q(partition__startswith=prefix) for prefix in filters],
             )
-            qs = qs.filter(filter_matches=True)
-            filt_maxes = qs.values("instance_id").annotate(maxval=Max("counter"))
-            per_filter_max.append(
-                {dmc["instance_id"]: dmc["maxval"] for dmc in filt_maxes}
+            sub_partitions = set(
+                queryset.filter(sub_condition)
+                .values_list("partition", flat=True)
+                .distinct()
             )
 
-        instance_id_lists = [maxes.keys() for maxes in per_filter_max]
-        all_instance_ids = reduce(set.union, instance_id_lists, set())
-        if is_producer:
-            # when we're sending, we want to make sure we include everything
-            result = {
-                instance_id: max([d.get(instance_id, 0) for d in per_filter_max])
-                for instance_id in all_instance_ids
+            # get the DMC records with partitions that are prefixes of the filters
+            super_partitions = set()
+            for filt in filters:
+                qs = (
+                    queryset.annotate(
+                        filter_matches=ValueStartsWithField(filt, "partition")
+                    )
+                    .filter(filter_matches=True)
+                    .exclude(partition=filt)
+                )
+                super_partitions.update(
+                    qs.values_list("partition", flat=True).distinct()
+                )
+
+            # get the instance counters for the partitions, filtered (if requested) by instance_ids still used in the store
+            super_fsics = {}
+            for part in super_partitions:
+                super_fsics[part] = cls.get_instance_counters_for_partition(
+                    part, only_if_in_store=only_if_in_store
+                )
+            sub_fsics = {}
+            for part in sub_partitions:
+                sub_fsics[part] = cls.get_instance_counters_for_partition(
+                    part, only_if_in_store=only_if_in_store
+                )
```

Comment:
For the above two super/sub steps that gather the instance counters while `only_if_in_store=True`, I think we can achieve a speedup by reducing them to a single query for each super/sub case. By structuring the query in a certain way, we can query for multiple partitions in a single query and that should scale significantly better than O(n), in contrast with this. 

The original code produces a query like:
```sql
SELECT morango_databasemaxcounter.instance_id, morango_databasemaxcounter.counter
FROM morango_databasemaxcounter
WHERE (
    morango_databasemaxcounter.partition = '5154da78c739fc2f678a9a549b1dc565'
    AND morango_databasemaxcounter.instance_id IN (
        SELECT DISTINCT U0.last_saved_instance AS Col1
        FROM morango_store U0
        WHERE U0.partition LIKE '5154da78c739fc2f678a9a549b1dc565%'
    )
);
```
and t takes roughly 17 seconds to execute on KDP. We should expect that to scale somewhat linearly with the number of partitions, so with a 60 second timeout we can query for ~3.5 partitions. If I understand correctly, this would be more of an issue with SoUDs and the "sub" partition case, since there would be distinct "partitions" to query for each SoUD?

An alternate approach to writing the same query uses `EXISTS` and we get similar performance:
```sql
SELECT morango_databasemaxcounter.instance_id, morango_databasemaxcounter.counter
FROM morango_databasemaxcounter
WHERE (
    morango_databasemaxcounter.partition = '5154da78c739fc2f678a9a549b1dc565'
    AND EXISTS(
        SELECT 1
        FROM morango_store U0
        WHERE (
            U0.partition LIKE morango_databasemaxcounter.partition || '%'
            AND U0.last_saved_instance = (morango_databasemaxcounter.instance_id)
        )
    )
);
```
This works in SQLite and Postgres. It's slightly slower in SQLite, but not enough to warrant attention.

With that change, we can further alter it to query for multiple at once:
```sql
SELECT morango_databasemaxcounter.partition, morango_databasemaxcounter.instance_id, morango_databasemaxcounter.counter
FROM morango_databasemaxcounter
WHERE (
    morango_databasemaxcounter.partition IN ('5154da78c739fc2f678a9a549b1dc565')
    AND EXISTS(                            
        SELECT 1                           
        FROM morango_store U0
        WHERE (
            U0.partition LIKE morango_databasemaxcounter.partition || '%'
            AND U0.last_saved_instance = (morango_databasemaxcounter.instance_id)
        )                    
    )          
);
```
The performance of that query slows to ~23 seconds, but querying with 2-3 partitions doesn't increase the query time, so when there are 3 partitions to query for, we get a 50% improvement over the original. With more partitions, we'll see even better performance in comparison. The downside is that there's a bug in Django that adds `= TRUE` to the `EXISTS(...)` clause when generating a query like this. It's fixed in later versions of Django but it causes postgres (9.6 at least) to choose poorer query plan even though it actually uses indices in that case.

Overall, I think we should avoid looping on this expensive query. Happy to help figure out how to make Django happy.

Reply:
Brilliant, thanks for the detailed analysis and benchmarking on this!

>If I understand correctly, this would be more of an issue with SoUDs and the "sub" partition case, since there would be distinct "partitions" to query for each SoUD?

That's correct -- there's typically one new subpartition added (with instance/counters specific to the SoUD, so they aren't redundant with a superpartition) per user who syncs from a SoUD. So parallelizing into a single query like that is definitely desirable.

>This works in SQLite and Postgres. It's slightly slower in SQLite, but not enough to warrant attention.

Just out of curiosity -- slower than Postgres, or also slower than the original query?

>Overall, I think we should avoid looping on this expensive query. Happy to help figure out how to make Django happy.

Thanks, would love your help with this!

A further thought: there may be even more ways to further consolidate, given that the list of partitions we're looping over there are themselves the products of applying the filters (of which there are generally very few) against the partitions in the DMC table. We'd still want a separate query for super vs sub, but other than that it should be possible. I kept things more separate for now given I was wanting to be able to easily introspect and reason about the inputs/outputs of each stage (and toggle on/off certain things), while finalizing the algorithm, but given it seems to be working, we could look towards further consolidation into some larger queries. Except if it doesn't buy us too much in terms of performance, it may not be worth it for now?

Reply:
> Just out of curiosity -- slower than Postgres, or also slower than the original query?

The original query took 2-3ms on SQLite for the VPN DB, and the query using `EXISTS` took 6-7ms. So it's twice as slow, but also 3-5ms isn't very much.

> Except if it doesn't buy us too much in terms of performance, it may not be worth it for now?

Since we're already dealing with timeouts on KDP, I think we should try to avoid making that worse. Some small tweaks to avoid repeating expensive queries are some quick wins we can make to mitigate it. Like you said, with tests in place and passing, it's easy to swap out bits of code and verify the algorithm isn't altered negatively. Here's what makes Django happy and gets around the `EXISTS(...) = TRUE` issue:

```python
    @classmethod
    def get_instance_counters_for_partitions(cls, partitions, only_if_in_store=True):
        queryset = cls.objects.filter(partition__in=partitions)
        if only_if_in_store:
            matching_store_instances = (
                Store.objects.filter(
                    partition__startswith=models.OuterRef("partition"),
                    last_saved_instance=models.OuterRef("instance_id")
                )
                .values("last_saved_instance")
            )
            # manually add EXISTS clause to avoid Django bug which puts `= TRUE` comparison that kills postgres perf
            exists = models.Exists(matching_store_instances)\
                .resolve_expression(query=queryset.query, allow_joins=True, reuse=None)
            queryset.query.where.add(exists, "AND")

        queryset = queryset.values_list("partition", "instance_id", "counter")

        counters = {}
        for partition, instance_id, counter in queryset:
            if partition not in counters:
                counters[partition] = {}
            counters[partition].update({instance_id: counter})
        return counters
```

Reply:
Thanks so much! I'll give that a shot -- there are a couple more regressions tests I'm thinking to add now as well (to cover some bases with the `only_if_in_store` piece), just to be more fully confident on having that as the default.

>Since we're already dealing with timeouts on KDP, I think we should try to avoid making that worse. Some small tweaks to avoid repeating expensive queries are some quick wins we can make to mitigate it.

Sorry, I wasn't being very clear: I very definitely agree that avoiding the duplicated queries as you're doing here is critical. I was referring there to my suggestion that we could go even further and build more of a mega-query that didn't require us to first query for the list of partitions and then pass them in to this function. But I think we can work towards that part incrementally -- this piece here is the critical lower-hanging fruit.